File: source/ts/src/core/samvat.ts
import type { SamvatInfo } from '../types/elements';
import { boundingNewMoons } from '../astronomy/newMoon';
import { getSiderealSunLongitude } from '../astronomy/sun';

const DAY_MS = 86_400_000;

const SAMVATSARA_NAMES: readonly string[] = [
  'Prabhava', 'Vibhava', 'Shukla', 'Pramoda', 'Prajapati', 'Angirasa',
  'Shrimukha', 'Bhava', 'Yuva', 'Dhata', 'Ishvara', 'Bahudhanya',
  'Pramathi', 'Vikrama', 'Vrisha', 'Chitrabhanu', 'Svabhanu', 'Tarana',
  'Parthiva', 'Vyaya', 'Sarvajit', 'Sarvadhari', 'Virodhi', 'Vikriti',
  'Khara', 'Nandana', 'Vijaya', 'Jaya', 'Manmatha', 'Durmukhi',
  'Hevilambi', 'Vilambi', 'Vikari', 'Sharvari', 'Plava', 'Shubhakrit',
  'Shobhakrit', 'Krodhi', 'Vishvavasu', 'Parabhava', 'Plavanga', 'Kilaka',
  'Saumya', 'Sadharana', 'Virodhakrit', 'Paridhavi', 'Pramadi', 'Ananda',
  'Rakshasa', 'Nala', 'Pingala', 'Kalayukta', 'Siddharthi', 'Raudra',
  'Durmati', 'Dundubhi', 'Rudhirodgari', 'Raktakshi', 'Krodhana', 'Akshaya',
];

const chaitraCache = new Map<number, number>();

/** The Chaitra new moon opening the lunar year: the last one before Mesha Sankranti. */
export function chaitraNewMoon(gregYear: number): number {
  const cached = chaitraCache.get(gregYear);
  if (cached !== undefined) return cached;

  let ref = new Date(Date.UTC(gregYear, 0, 20));
  let result = Date.UTC(gregYear, 2, 22); // fallback; should not be hit
  for (let i = 0; i < 6; i++) {
    const { next } = boundingNewMoons(ref);
    const sun = getSiderealSunLongitude(next, 'lahiri');
    if (sun >= 330 && sun < 360) {
      result = next.getTime();
      break;
    }
    ref = new Date(next.getTime() + DAY_MS);
  }
  chaitraCache.set(gregYear, result);
  return result;
}

/**
 * Both eras turn over at Chaitra Shukla Pratipada; the +9 / +11 name offsets encode
 * the 13-position gap between the northern and southern samvatsara lists.
 */
export function computeSamvat(date: Date): SamvatInfo {
  const gregYear = date.getUTCFullYear();
  const pastNewYear = date.getTime() >= chaitraNewMoon(gregYear);

  const vikramSamvat = gregYear + (pastNewYear ? 57 : 56);
  const shakaSamvat = gregYear + (pastNewYear ? -78 : -79);

  return {
    vikramSamvat,
    shakaSamvat,
    vikramSamvatsara: SAMVATSARA_NAMES[(((vikramSamvat + 9) % 60) + 60) % 60]!,
    shakaSamvatsara: SAMVATSARA_NAMES[(((shakaSamvat + 11) % 60) + 60) % 60]!,
  };
}

File: source/go/internal/core/samvat.go
package core

import (
	"github.com/ishankgupta95/panchang/source/go/v5/internal/astronomy"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/store"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

const samvatDayMs = 86_400_000

var samvatsaraNames = [60]string{
	"Prabhava", "Vibhava", "Shukla", "Pramoda", "Prajapati", "Angirasa",
	"Shrimukha", "Bhava", "Yuva", "Dhata", "Ishvara", "Bahudhanya",
	"Pramathi", "Vikrama", "Vrisha", "Chitrabhanu", "Svabhanu", "Tarana",
	"Parthiva", "Vyaya", "Sarvajit", "Sarvadhari", "Virodhi", "Vikriti",
	"Khara", "Nandana", "Vijaya", "Jaya", "Manmatha", "Durmukhi",
	"Hevilambi", "Vilambi", "Vikari", "Sharvari", "Plava", "Shubhakrit",
	"Shobhakrit", "Krodhi", "Vishvavasu", "Parabhava", "Plavanga", "Kilaka",
	"Saumya", "Sadharana", "Virodhakrit", "Paridhavi", "Pramadi", "Ananda",
	"Rakshasa", "Nala", "Pingala", "Kalayukta", "Siddharthi", "Raudra",
	"Durmati", "Dundubhi", "Rudhirodgari", "Raktakshi", "Krodhana", "Akshaya",
}

const chaitraCacheYears = 512

var chaitraStore = store.New[int64, int64](chaitraCacheYears, store.DefaultStripes, store.HashInt64)

func ChaitraNewMoon(ctx *astronomy.EphemerisCtx, gregYear int) (int64, error) {
	key := int64(gregYear)
	if cached, ok := chaitraStore.Get(key); ok {
		return cached, nil
	}

	ref := utils.UtcDateMs(gregYear, 0, 20)
	result := utils.UtcDateMs(gregYear, 2, 22)
	for i := 0; i < 6; i++ {
		bounds, err := astronomy.BoundingNewMoons(ctx, ref)
		if err != nil {
			return 0, err
		}
		sun, err := astronomy.GetSiderealSunLongitude(ctx, bounds.NextMs, types.Lahiri)
		if err != nil {
			return 0, err
		}
		if sun >= 330 && sun < 360 {
			result = bounds.NextMs
			break
		}
		ref = bounds.NextMs + samvatDayMs
	}
	chaitraStore.Put(key, result)
	return result, nil
}

func ComputeSamvat(ctx *astronomy.EphemerisCtx, dateMs int64) (types.SamvatInfo, error) {
	gregYear := types.Date(dateMs).UTCFullYear()
	chaitra, err := ChaitraNewMoon(ctx, gregYear)
	if err != nil {
		return types.SamvatInfo{}, err
	}
	pastNewYear := dateMs >= chaitra

	vikramSamvat := gregYear + 56
	shakaSamvat := gregYear - 79
	if pastNewYear {
		vikramSamvat = gregYear + 57
		shakaSamvat = gregYear - 78
	}

	return types.SamvatInfo{
		VikramSamvat:     vikramSamvat,
		ShakaSamvat:      shakaSamvat,
		VikramSamvatsara: samvatsaraNames[samvatsaraIndex(vikramSamvat+9)],
		ShakaSamvatsara:  samvatsaraNames[samvatsaraIndex(shakaSamvat+11)],
	}, nil
}

func samvatsaraIndex(n int) int {
	return ((n % 60) + 60) % 60
}

func ClearChaitraCache() { chaitraStore.Clear() }
